File: src/lib/conciliacao/conciliacaoCalc.ts
export const QUITADA_TOLERANCE = 0.01;

export type ConciliacaoTotais = {
  valorAcao: number;
  valorNotasVinculadas: number;
  valorPagoTotal: number;
  saldoEmAberto: number;
  percentualPago: number;
  quitada: boolean;
  qtdNotas: number;
};

export function calcConciliacaoTotais(
  valorAcao: number,
  valorNotasVinculadas: number,
  valorPagoTotal: number,
  qtdNotas = 0,
): ConciliacaoTotais {
  const saldoEmAberto = Math.max(0, valorAcao - valorPagoTotal);
  const percentualPago = valorAcao > 0 ? (valorPagoTotal / valorAcao) * 100 : 0;
  return {
    valorAcao,
    valorNotasVinculadas,
    valorPagoTotal,
    saldoEmAberto,
    percentualPago,
    quitada: valorPagoTotal >= valorAcao - QUITADA_TOLERANCE,
    qtdNotas,
  };
}

export function deriveNotaFiscalStatus(valorServico: number, valorPago: number): string {
  if (valorPago >= valorServico - QUITADA_TOLERANCE) {
    return 'Pago';
  }
  if (valorPago > 0) {
    return 'Pago Parcial';
  }
  return 'Pendente';
}

export function formatBRL(value: number): string {
  return new Intl.NumberFormat('pt-BR', { style: 'currency', currency: 'BRL' }).format(value);
}

File: backend/internal/domain/service/conciliacao_calc.go
package service

import "math"

const ConciliacaoQuitadaTolerance = 0.01

// ConciliacaoTotais métricas derivadas para uma ação judicial.
type ConciliacaoTotais struct {
	ValorNotasVinculadas float64
	ValorPagoTotal       float64
	SaldoEmAberto        float64
	PercentualPago       float64
	Quitada              bool
}

func CalcConciliacaoTotais(valorAcao, valorNotasVinculadas, valorPagoTotal float64) ConciliacaoTotais {
	saldo := valorAcao - valorPagoTotal
	if saldo < 0 {
		saldo = 0
	}
	pct := 0.0
	if valorAcao > 0 {
		pct = (valorPagoTotal / valorAcao) * 100
	}
	return ConciliacaoTotais{
		ValorNotasVinculadas: valorNotasVinculadas,
		ValorPagoTotal:       valorPagoTotal,
		SaldoEmAberto:        saldo,
		PercentualPago:       pct,
		Quitada:              valorPagoTotal >= valorAcao-ConciliacaoQuitadaTolerance,
	}
}

func DeriveNotaFiscalStatus(valorServico, valorPago float64) string {
	if valorPago >= valorServico-ConciliacaoQuitadaTolerance {
		return "Pago"
	}
	if valorPago > 0 {
		return "Pago Parcial"
	}
	return "Pendente"
}

func roundMoney(v float64) float64 {
	return math.Round(v*100) / 100
}
